src/twisted/internet/test/test_fdset.py

```diff
@@ -115,6 +115,19 @@ def fail():
         # Getting here means the right thing happened probably.
 
 
+    def test_removeReader_selectables_nokey(self):
+        """
+        L{reactor.removeReader()} accepts an L{IReadDescriptor} provider
+        previously passed to C{reactor.addReader()}. Checks if it doesn't
+        raise KeyError exception
+        """
+        reactor, fd, server = self._simpleSetup()
+        reactor.addReader(fd)
+        reactor._selectables = {}
```

Comment:
Thanks for the contribution @DefaultUser !  Having a reproducable example of this problem makes it much easier to understand what's wrong.

In the case of the linked rpy script, the problem looks like it's in the application code.  It uses `deferToThread` to schedule a function to run in a non-reactor thread.  Then, that function calls a Twisted API directly (`request.write` and `request.finish`).  This isn't supported and broken behavior as a result is just what I'd expect.  Instead, the Twisted APIs need to be invoked by calling *back* into the reactor thread.  Something like:

    def _render_thread(self, request):
        time.sleep(1)
        def do_it():
            request.write("<html><body>something from a thread</body></html>")
            request.finish()
        reactor.callInThread(do_it)

(Not that this is an example of _good_ multithreaded coded - it's almost certainly an example of _bad_ multithreaded code, in fact.  But this is the smallest possible change to the example code to make it use the Twisted APIs correctly.)

Again, thanks for sharing this example.  It helps to know what kind of behavior is possible from this misuse of the Twisted APIs (this might help other people diagnose such misuses).  In fact, perhaps the original reporter will report that this is analogous to his or her circumstances and then we'll finally understand the original problem report.  I hope so.